### PR TITLE
feat: rewrite defineObjectProp logic

### PR DESCRIPTION
**Pull request by BQXBQX**
##### Checklist

- [x] `npm test` passes
image

- [x] commit message follows commit guidelines

##### Description of change

- Refactored logic to handle `null` and `undefined` parameters for proper storage.
- Used `Object.assign` to merge objects without altering the outer reference.
- Improved branching to handle different argument types clearly.
- Enhanced code readability and extensibility.
- Added 6551 test cases to ensure thorough validation of edge cases.
- Fixed https://github.com/antvis/G2/issues/6551

### CODE REVIEW

File: src/api/define.ts [hunk 1]
```diff
@@ -39,13 +39,36 @@ function defineObjectProp(
...
+        if (typeof key !== 'string') {
+          return this.attr(k, key);
+        }
+
+        currentObj[key] = true;
+        return this.attr(k, currentObj);
+
+      case 2:
+        currentObj[key] = value;
+        return this.attr(k, currentObj);
```
Comment (hustcc) [inline src/api/define.ts:67]: 感觉写法也有点复杂了😵‍💫

Comment (BQXBQX) [reply src/api/define.ts:67]: 是逻辑上的精简还是写法上的精简呢？感觉逻辑上无法优化了，逻辑上再优化单测就出问题了 😵‍💫

Comment (hustcc) [reply src/api/define.ts:67]: 逻辑省不了，就是代码上了～

**Comment by hustcc** [comment]
@BQXBQX 仔细想了一下，感觉不是很好规整。当前 `defineObjectProp ` 有几种用法，分别是 单个设置、批量设置，这个 PR 的改动，是让 `defineObjectProp` 支持批量设置的 merge，可能带来几个问题：

1. 代码上可能也更复杂了，并不比之前更易读，毕竟代码上功能更复杂了。

2. 批量设置的用法上，针对 object、boolean、number、string 的逻辑其实不一致的，虽然是基本符合用户认知，但是前提是 defineObjectProp 的 API 都是 key value 形式的，也就是类似于 interaction、encode 这类，当前是符合的，但是担心未来可能有其他的形式，需要预留一些余量，比如：

```ts
chart.newAPI({ globalXXX: true })
```

3. 做完 merge 操作之后，可能也会有另外的需求，就是是否需要 deepMerge？deepMerge 有会遇到数字的 merge 问题。其实有一种做不完的感觉。


所以建议：

1. 维持现在的功能，object 的时候，就是覆盖。
2. 尝试简化现有代码中 arguments.length 的逻辑
3. 在文档中，标注出来两种用法和作用，告知 传入 object 就是全覆盖的操作，而非 megre。

你看看觉得如何？

**Comment by BQXBQX** [comment]
> @BQXBQX 仔细想了一下，感觉不是很好规整。当前 `defineObjectProp ` 有几种用法，分别是 单个设置、批量设置，这个 PR 的改动，是让 `defineObjectProp` 支持批量设置的 merge，可能带来几个问题：
> 
> 1. 代码上可能也更复杂了，并不比之前更易读，毕竟代码上功能更复杂了。
> 2. 批量设置的用法上，针对 object、boolean、number、string 的逻辑其实不一致的，虽然是基本符合用户认知，但是前提是 defineObjectProp 的 API 都是 key value 形式的，也就是类似于 interaction、encode 这类，当前是符合的，但是担心未来可能有其他的形式，需要预留一些余量，比如：
> 
> ```ts
> chart.newAPI({ globalXXX: true })
> ```
> 
> 3. 做完 merge 操作之后，可能也会有另外的需求，就是是否需要 deepMerge？deepMerge 有会遇到数字的 merge 问题。其实有一种做不完的感觉。
> 
> 所以建议：
> 
> 1. 维持现在的功能，object 的时候，就是覆盖。
> 2. 尝试简化现有代码中 arguments.length 的逻辑
> 3. 在文档中，标注出来两种用法和作用，告知 传入 object 就是全覆盖的操作，而非 megre。
> 
> 你看看觉得如何？

好的，那我把文档说明加在 `FAQ` 里面可以吗？

**Comment by hustcc** [comment]
需要在 [这里](https://g2.antv.antgroup.com/api/mark#markinteraction) 写的，这个文档目前都还没有 interaction 的 API。

**Comment by BQXBQX** [comment]
> 需要在 [这里](https://g2.antv.antgroup.com/api/mark#markinteraction) 写的，这个文档目前都还没有 interaction 的 API。
> 
> ```ts
> export const commonProps = {
>   encode: { type: 'object' },
>   scale: { type: 'object' },
>   data: { type: 'value' },
>   transform: { type: 'array' },
>   style: { type: 'object' },
>   animate: { type: 'object' },
>   coordinate: { type: 'object' },
>   interaction: { type: 'object' },
>   label: { type: 'array', key: 'labels' },
>   axis: { type: 'object' },
>   legend: { type: 'object' },
>   slider: { type: 'object' },
>   scrollbar: { type: 'object' },
>   state: { type: 'object' },
>   layout: { type: 'object' },
>   theme: { type: 'object' },
>   title: { type: 'value' },
> } as const;
> ```
> 
> 理论上，上面 type 是 object 的都是走的这个逻辑。

要是 type 是 object 都会出现这种情况，把这个问题放到 interaction 是不是有点不合适 🤕

**Comment by hustcc** [comment]
感觉冗余点，都要加，这个就是文档难搞的一个点，跟着代码有区别。

**Comment by hustcc** [comment]
要不这个先 hold 吧，明年我们做文档优化，一起来处理。
